StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c

```diff
@@ -0,0 +1,800 @@
+             PlatformHobSize,
+             MmFvBase,
+             MmFvSize,
+             MmCoreFileName,
+             MmCoreImageAddress,
+             MmCoreImageSize,
+             MmCoreEntryPoint,
+             MmProfileDataHob,
+             Block
+             );
+  ASSERT_EFI_ERROR (Status);
+
+  //
+  // Create MM HOB list end.
+  //
+  MmIplCreateHob ((UINT8 *)HobList + PlatformHobSize + FoundationHobSize, EFI_HOB_TYPE_END_OF_HOB_LIST, sizeof (EFI_HOB_GENERIC_HEADER));
+
+  return HobList;
+}
+
+/**
+  Find largest unallocated MMRAM in current MMRAM descriptor block
+
+  @param[in, out] LagestMmramRangeIndex  Lagest mmram range index.
+  @param[in]      CurrentBlock           Current MMRAM descriptor block.
+
+**/
+VOID
+FindLargestMmramRange (
+  IN OUT UINTN                       *LagestMmramRangeIndex,
+  IN EFI_MMRAM_HOB_DESCRIPTOR_BLOCK  *CurrentBlock
+  )
+{
+  UINTN                 Index;
+  UINT64                MaxSize;
+  BOOLEAN               Found;
+  EFI_MMRAM_DESCRIPTOR  *MmramRanges;
+
+  MmramRanges = CurrentBlock->Descriptor;
+
+  //
+  // Find largest Mmram range.
+  //
+  Found = FALSE;
+  for (Index = 0, MaxSize = SIZE_256KB - EFI_PAGE_SIZE; Index < CurrentBlock->NumberOfMmReservedRegions; Index++) {
+    //
+    // Skip any MMRAM region that is already allocated, needs testing, or needs ECC initialization
+    //
+    if ((MmramRanges[Index].RegionState & (EFI_ALLOCATED | EFI_NEEDS_TESTING | EFI_NEEDS_ECC_INITIALIZATION)) != 0) {
+      continue;
+    }
+
+    if (MmramRanges[Index].CpuStart >= BASE_1MB) {
+      if ((MmramRanges[Index].CpuStart + MmramRanges[Index].PhysicalSize) <= BASE_4GB) {
+        if (MmramRanges[Index].PhysicalSize >= MaxSize) {
+          Found                  = TRUE;
+          *LagestMmramRangeIndex = Index;
+          MaxSize                = MmramRanges[Index].PhysicalSize;
+        }
+      }
+    }
+  }
+
+  if (Found == FALSE) {
+    DEBUG ((DEBUG_ERROR, "Not found largest unlocated MMRAM\n"));
+    ASSERT (FALSE);
+    CpuDeadLoop ();
+  }
+
+  return;
+}
+
+/**
+  Allocate available MMRAM for MM core image.
+
+  @param[in]  Pages                     Page count of MM core image.
+  @param[out] NewBlock                  Pointer of new mmram block HOB.
+
+  @return  EFI_PHYSICAL_ADDRESS         Address for MM core image to be loaded in MMRAM.
+**/
+EFI_PHYSICAL_ADDRESS
+MmIplAllocateMmramPage (
+  IN  UINTN                           Pages,
+  OUT EFI_MMRAM_HOB_DESCRIPTOR_BLOCK  **NewBlock
+  )
+{
+  UINTN                           LagestMmramRangeIndex;
+  UINT32                          FullMmramRangeCount;
+  EFI_HOB_GUID_TYPE               *MmramInfoHob;
+  EFI_MMRAM_DESCRIPTOR            *Largest;
+  EFI_MMRAM_DESCRIPTOR            *Allocated;
+  EFI_MMRAM_DESCRIPTOR            *FullMmramRanges;
+  EFI_MMRAM_HOB_DESCRIPTOR_BLOCK  *CurrentBlock;
+  EFI_MMRAM_HOB_DESCRIPTOR_BLOCK  *NewDescriptorBlock;
+
+  MmramInfoHob = GetFirstGuidHob (&gEfiSmmSmramMemoryGuid);

```

Comment:
It seems that this change is sticking to the tradition of SmmRelocationLib to grab the MMRAM regions from the hobs. But StandaloneMmCore is pulling this information from different sources: https://github.com/tianocore/edk2/blob/a29a9cce5f9afa32560d966e501247246ec96ef6/StandaloneMmPkg/Core/StandaloneMmCore.c#L540. Is there any plan to consolidate the usage?

Reply:
Also, the tradition SMM IPL is using a combination of MM configuration protocol and MM Access protocol to figure out the ranges. If the platform has the PPI versions of these 2 interfaces available, which source should the IPL follow, if any discrepancies present?